HBASE-27178 Use spotless to format code (including scala code)

NihalJain opened the pull request:
- Added the base plugin to pom.xml (similar to other hbase repos)
- Extended the plugin to support formatting of scala code / headers
- Added the files required by the base plugin ( copying them from hbase repo)
- Added a new file .scalafmt.conf, which is required by spotless for scala (copied from spark repo)
- Made changes to fix docstring param new line issue and used latest scalafmt version i.e. 3.7.12

NihalJain commented:
Commit 1: Added plug-in along with required files (including support for scala code)
Commit 2:
Ran mvn spotless:apply and staged all file changes. We can decide if we want to do this as a separate change to keep things clean.

Let me know what do you all think of this change.

CC: @duo @ndimiduk @wchevreuil @petersomogyi @Reidddddd

NihalJain commented on spark/hbase-spark/src/main/java/org/apache/hadoop/hbase/spark/example/hbasecontext/JavaHBaseBulkLoadExample.java:
```diff
@@ -35,19 +36,17 @@
 import org.apache.yetus.audience.InterfaceAudience;
 
 /**
- * Run this example using command below:
```
wrap in code block

Reidddddd commented:
> Commit 1: Added plug-in along with required files (including support for scala code) Commit 2: Ran mvn spotless:apply and staged all file changes. We can decide if we want to do this as a separate change to keep things clean.
> 
> Let me know what do you all think of this change.
> 
> CC: @duo @ndimiduk @wchevreuil @petersomogyi @Reidddddd

duo is this account: @Apache9

Apache-HBase commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 26s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  1s |  markdownlint was not available.  |
| +0 :ok: |  shelldocs  |   0m  1s |  Shelldocs was not available.  |
| +0 :ok: |  spotbugs  |   0m  0s |  spotbugs executables are not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 20 new or modified test files.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 27s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   1m 45s |  master passed  |
| +1 :green_heart: |  compile  |   1m  1s |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 57s |  master passed  |
| +1 :green_heart: |  scaladoc  |   0m  0s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m  7s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   0m 53s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 12s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 12s |  the patch passed  |
| +1 :green_heart: |  scalac  |   1m 12s |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  0s |  There were no new shellcheck issues.  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  xml  |   0m 10s |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  javadoc  |   1m  0s |  the patch passed  |
| +1 :green_heart: |  scaladoc  |   0m  0s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   9m  3s |  root in the patch passed.  |
|  |   |  19m 29s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-119/1/artifact/yetus-precommit-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase-connectors/pull/119 |
| Optional Tests | dupname markdownlint xml shellcheck shelldocs javac javadoc unit compile spotbugs findbugs scalac scaladoc |
| uname | Linux ff88f39620be 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 GNU/Linux |
| Build tool | hb_maven |
| Personality | dev-support/jenkins/hbase-personality.sh |
| git revision | master / bfcdf68 |
| Default Java | Oracle Corporation-1.8.0_282-b08 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-119/1/testReport/ |
| Max. process+thread count | 954 (vs. ulimit of 12500) |
| modules | C: kafka kafka/hbase-kafka-model kafka/hbase-kafka-proxy spark spark/hbase-spark-protocol spark/hbase-spark-protocol-shaded spark/hbase-spark spark/hbase-spark-it hbase-connectors-assembly . test-reporting U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-119/1/console |
| versions | git=2.20.1 shellcheck=0.5.0 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

Reidddddd commented:
>> if we want to do this as a separate change to keep things clean.

I'd like the separation

NihalJain commented:
> > > if we want to do this as a separate change to keep things clean.
> 
> I'd like the separation

Sure will force push and drop 2nd commit once it looks good to others

NihalJain commented on spark/hbase-spark/src/main/scala/org/apache/hadoop/hbase/spark/datasources/JavaBytesEncoder.scala:
```diff
@@ -25,84 +25,101 @@ import org.apache.yetus.audience.InterfaceStability
+   * The function performing real filtering operations. The format of filterBytes depends on the
+   * implementation of the BytesEncoder.
+   *
+   * @param input:
```
its an issue with scalafmt plugin version we are using: https://github.com/scalameta/scalafmt/issues/1387
Will try to bump up to fixed version and update here: https://github.com/scalameta/scalafmt/pull/1987 or apply workaround (as last resort)

NihalJain replied:
update: bumping did not work, so had to apply workaround

NihalJain commented:
Reverted 2nd commit and fix scala param issue. The changes 

> Commit 1: Added plug-in along with required files (including support for scala code) Commit 2: Ran mvn spotless:apply and staged all file changes. We can decide if we want to do this as a separate change to keep things clean.
> 
> Let me know what do you all think of this change.
> 
> CC: @duo @ndimiduk @wchevreuil @petersomogyi @Reidddddd

Dropped 2nd commit as discuused with @Reidddddd. Will raise a WIP-PR with results after applying commit 1 here and running 'mvn spotless:apply'.

wForget commented on dev-support/.scalafmt.conf:
```diff
@@ -0,0 +1,38 @@
+    runner.dialect = scala213
+  }
+}
+version = 3.7.12
```
nit: end with newline

NihalJain replied:
fixed

Apache-HBase commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 35s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  shelldocs  |   0m  0s |  Shelldocs was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -0 :warning: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 40s |  master passed  |
| +1 :green_heart: |  compile  |   1m 16s |  master passed  |
| +1 :green_heart: |  javadoc  |   1m 28s |  master passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m  5s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 11s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 11s |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  0s |  There were no new shellcheck issues.  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  xml  |   0m  2s |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  javadoc  |   1m  5s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   9m 28s |  root in the patch passed.  |
|  |   |  19m 58s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-119/2/artifact/yetus-precommit-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase-connectors/pull/119 |
| Optional Tests | dupname xml shellcheck shelldocs javac javadoc unit compile |
| uname | Linux 6dbfd1e806e7 5.4.0-152-generic #169-Ubuntu SMP Tue Jun 6 22:23:09 UTC 2023 x86_64 GNU/Linux |
| Build tool | hb_maven |
| Personality | dev-support/jenkins/hbase-personality.sh |
| git revision | master / bfcdf68 |
| Default Java | Oracle Corporation-1.8.0_282-b08 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-119/2/testReport/ |
| Max. process+thread count | 959 (vs. ulimit of 12500) |
| modules | C: . U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-119/2/console |
| versions | git=2.20.1 shellcheck=0.5.0 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

NihalJain commented on dev-support/.scalafmt.conf:
```diff
@@ -0,0 +1,38 @@
+docstrings.style = Asterisk
+# See https://github.com/scalameta/scalafmt/issues/1387
+docstrings.wrap = no
+maxColumn = 98
```
Need feedback: should we make this 100, similar to setting that we have for java? Or keep it same as spark project?

Reidddddd replied:
let's align with hbase project

NihalJain commented on dev-support/.scalafmt.conf:
```diff
@@ -0,0 +1,38 @@
+docstrings.wrap = no
+maxColumn = 98
+runner.dialect = scala212
+fileOverride {
```
Need feedback: IMHO we can drop this as we do not have this directory/implementation as spark has: https://github.com/apache/spark/tree/master/core/src/main/scala-2.13

Reidddddd replied:
+1

NihalJain replied:
fixed in commit 2

Apache-HBase commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 29s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  shelldocs  |   0m  0s |  Shelldocs was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -0 :warning: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 33s |  master passed  |
| +1 :green_heart: |  compile  |   0m 54s |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 51s |  master passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 49s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 52s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 52s |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  0s |  There were no new shellcheck issues.  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  xml  |   0m  2s |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  javadoc  |   0m 51s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   8m 56s |  root in the patch passed.  |
|  |   |  17m 15s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-119/3/artifact/yetus-precommit-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase-connectors/pull/119 |
| Optional Tests | dupname xml shellcheck shelldocs javac javadoc unit compile |
| uname | Linux 0152dd29d01b 5.4.0-152-generic #169-Ubuntu SMP Tue Jun 6 22:23:09 UTC 2023 x86_64 GNU/Linux |
| Build tool | hb_maven |
| Personality | dev-support/jenkins/hbase-personality.sh |
| git revision | master / bfcdf68 |
| Default Java | Oracle Corporation-1.8.0_282-b08 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-119/3/testReport/ |
| Max. process+thread count | 959 (vs. ulimit of 12500) |
| modules | C: . U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-Connectors-PreCommit/job/PR-119/3/console |
| versions | git=2.20.1 shellcheck=0.5.0 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

Reidddddd commented:
let's wait some time whether there are other comments

NihalJain commented:
Thanks for review / merge. @Reidddddd / @wForget 
I have rebased https://github.com/apache/hbase-connectors/pull/120 with this change.
